### PR TITLE
FCP X XML adapter

### PR DESCRIPTION
**Pull request by eric-with-a-c**
This is a basic adapter for working with FCP X XMLs. While this adapter will handle basic FCP X projects, including compound clips, it does not, at present, support markers or video clips that have detached audio. 

ffprobe is required to create FCP X XMLs from an otio without errors on import. If ffprobe is not available or is not in the path, then an error will be displayed in FCP X when importing the XML. However, the timeline will still look an function properly in FCP X. 

#37

### CODE REVIEW

**Comment by jminor** [comment]
Thanks @eric-with-a-c ! This is really helpful. We'll take a closer look when we get back from SIGGRAPH.

File: opentimelineio_contrib/adapters/fcpx_xml.py [hunk 1]
```diff
@@ -0,0 +1,834 @@
...
+        str: The format name. If empty, then ffprobe couldn't find the item
+    """
+
+    path = path.replace("file://", "")
+    path = unquote(path)
+    if not os.path.exists(path):
+        return ""
+
+    try:
+        frame_size = subprocess.check_output(
```
[the remark this answers is not in the document]
Comment (eric-with-a-c) [reply opentimelineio_contrib/adapters/fcpx_xml.py]: Would using the warnings module be ok, or would you just want a print statement?

File: opentimelineio_contrib/adapters/tests/test_fcpx_adapter.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    The test class for the FCP X XML adapter
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(AdaptersFcpXXmlTest, self).__init__(*args, **kwargs)
+        self.maxDiff = None
+
+    def test_roundtrip(self):
+        timeline = otio.adapters.read_from_file(SAMPLE_XML)
+        self.assertTrue(timeline is not None)
```
[the remark this answers is not in the document]
Comment (eric-with-a-c) [reply opentimelineio_contrib/adapters/tests/test_fcpx_adapter.py]: Thanks for catching that! It should be fixed now.

File: opentimelineio_contrib/adapters/tests/test_fcpx_adapter.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        )
+
+        for n, track in enumerate(video_tracks):
+            self.assertTupleEqual(
+                tuple(c.name for c in track),
+                video_clip_names[n]
+            )
+
+        fcpx_xml = otio.adapters.write_to_string(timeline, "fcpx_xml")
+        self.assertTrue(fcpx_xml is not None)
```
[the remark this answers is not in the document]
Comment (eric-with-a-c) [reply opentimelineio_contrib/adapters/tests/test_fcpx_adapter.py]: This is also addressed

File: opentimelineio_contrib/adapters/fcpx_xml.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ssteinbach) [reply opentimelineio_contrib/adapters/fcpx_xml.py]: Good question!  We don't have a policy for adapter warnings or errors (see #28) so I think thats fine for now.  We'll have to go through these again later once we figure that out.  Thanks!

**Comment by jminor** [comment]
So we were able to track down an expert on the FCP X XML format. Here is a summary of their comments & advice:

- Since an FCP library can contain multiple sequences, the adapter could use a SerializableCollection of Timelines to hold several of them in one file.
- You might want to keep the event name in some metadata so you can reconstruct the organization if you round trip back to FCP.
- Using the lane position to build tracks makes sense.

Regarding the format information:
- It should be possible to include assets without a format, but you need to provide one for a sequence.
- There is a default of 720p60 for assets without a format.

**Comment by eric-with-a-c** [comment]
Thanks for the feedback!

I added the use of a SerializableCollection of Timelines to account for the possibility of having more than 1 sequence in an event. I also played around with adding another SerializableCollection of SerializableCollections since an FCP X Library can have multiple events, which can have multiple sequences. However, otioview didn't seem to like that level of nesting, so I left it out for now.

Presently the event name is being stored as the SerializableCollection's name and then the SerailizableCollection's name is used to set the event name when round tripping. But I could make it a piece of metadata if that's the preferred way of handling that. Let me know if you'd like it changed.

That's good to know about the formats. Since we still need a format for the sequence, it looks like there's still a dependency on ffprobe.

Let me know if there is anything else that needs to be adjusted before merging. 

Thanks again!

**Comment by jminor** [comment]
@eric-with-a-c, thanks for the contribution!

**Comment by latenitefilms** [comment]
Amazing work @eric-with-a-c ! Thank you!!

**Comment by apetrynet** [comment]
Hi @eric-with-a-c!

Sorry for reaching out to you via this old issue. I'm not sure if you're on the  ASWF slack(?).

I'm reaching out on the behalf of OpenTimelineIO as we're in the process of moving adapters out of the main OTIO repo and over to individual repos under OpenTimelineIO [organisation](https://github.com/OpenTimelineIO). We're doing this as part of the v1.0 release plan. 
Moving the adapters out of the core repo will ease maintenance/updates as they no longer are held back by the OTIO core release cycle. 
Since you're the main author of the "fcpx_xml" I'd like to ask if you'd prefer hosting this adapter under another location or if we should move it over to the OTIO organisation together with the other adapters?

Please let me know what you think.
You can reach me directly on the ASWF slack under my real name or in the "opentimelineio" channel.

**Pull request merged**